src/utils/tokenCounter.ts

```diff
@@ -0,0 +1,125 @@
+import { get_encoding, Tiktoken } from "tiktoken"
+
+/**
+ * Shared token counter utility using tiktoken's cl100k_base encoding
+ * This provides a universal proxy token counter that works reasonably well for most models
+ */
+class TokenCounter {
+	private encoding: Tiktoken | null = null
+	private initializationPromise: Promise<void> | null = null
+
+	/**
+	 * Initialize the tiktoken encoding (lazy loading)
+	 */
+	private async initialize(): Promise<void> {
+		if (this.encoding) return
+
+		if (!this.initializationPromise) {
+			this.initializationPromise = this.doInitialize()
+		}
+		return this.initializationPromise
+	}
+
+	private async doInitialize(): Promise<void> {
+		try {
+			this.encoding = get_encoding("cl100k_base")
+		} catch (error) {
+			console.warn("Failed to initialize tiktoken encoding:", error)
+			this.encoding = null
+		}
+	}
+
+	/**
+	 * Estimate token count for a text string using tiktoken
+	 * Falls back to character-based estimation if tiktoken fails
+	 */
+	async estimateTokens(text: string): Promise<number> {
+		if (!text) return 0
+
+		try {
+			await this.initialize()
+			if (this.encoding) {
+				return this.encoding.encode(text).length
+			}
+		} catch (error) {
+			console.warn("Tiktoken encoding failed, falling back to character estimation:", error)
+		}
+
+		// Fallback to character-based estimation (roughly 4 chars per token)
+		return Math.ceil(text.length / 4)
+	}
+
+	/**
+	 * Estimate token count synchronously using character-based approximation
+	 * Use this when you need immediate results without async overhead
+	 */
+	estimateTokensSync(text: string): number {
+		if (!text) return 0
+
+		// Try tiktoken if already initialized
+		if (this.encoding) {
+			try {
+				return this.encoding.encode(text).length
+			} catch (error) {
+				console.warn("Tiktoken encoding failed, falling back to character estimation:", error)
+			}
+		}
+
+		// Fallback to character-based estimation (roughly 4 chars per token)
+		return Math.ceil(text.length / 4)
+	}
+
+	/**
+	 * Extract text content from various message formats and estimate tokens
+	 */
+	async estimateTokensFromMessages(messages: any[]): Promise<number> {
```

Comment:
This looks unreliable as it only looks for text string but missing tools calls and other content type that also consume token?